### PR TITLE
Add withfield bytecode support

### PR DESCRIPTION
**Pull request by tajila**
Add support for the withfield bytecode

Based on Eric's changes in 
#1253

Signed-off-by: tajila <atobia@ca.ibm.com>

### CODE REVIEW

**Comment by tajila** [comment]
@gacholio Please review this PR

File: runtime/bcutil/ConstantPoolMap.hpp [hunk 1]
```diff
@@ -275,6 +275,7 @@ class ConstantPoolMap
 	void markClassAsUsedByNew(U_16 classCfrCPIndex)            { mark(classCfrCPIndex, NEW); }
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 	void markClassAsUsedByDefaultValue(U_16 classCfrCPIndex)   { mark(classCfrCPIndex, DEFAULT_VALUE); }
+	void markFieldRefAsUsedByWithField(U_16 fieldRefCfrCPIndex){ mark(fieldRefCfrCPIndex, WITH_FIELD); }
```
Comment (gacholio) [inline runtime/bcutil/ConstantPoolMap.hpp]: No space before {

File: runtime/oti/j9consts.h [hunk 2]
```diff
@@ -250,6 +250,9 @@ extern "C" {
 #define J9_RESOLVE_FLAG_JCL_CONSTANT_POOL 0x100
 #define J9_RESOLVE_FLAG_REDEFINE_CLASS 0x200
 #define J9_RESOLVE_FLAG_NO_CLASS_INIT 0x400
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+#define J9_RESOLVE_FLAG_FIELD_WITHER 0x800
```
Comment (gacholio) [inline runtime/oti/j9consts.h]: That's a very odd name.

File: runtime/vm/BytecodeInterpreter.hpp [hunk 3]
```diff
@@ -8114,8 +8114,91 @@ done:;
 	VMINLINE VM_BytecodeAction
 	withfield(REGISTER_ARGS_LIST)
 	{
+retry:
 		VM_BytecodeAction rc = EXECUTE_BYTECODE;
-		// TODO: Implement bytecode
+		const U_16 index = *(U_16 *)(_pc + 1);
```
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp]: Please put const after the type (here and elsewhere), rather than before.

File: runtime/vm/BytecodeInterpreter.hpp [hunk 4]
```diff
@@ -8114,8 +8114,91 @@ done:;
...
+		/* Must be done after resolution as we don't know where the object ref is
+		 * until we have resolved the fieldref
+		 */
+		if (NULL == originalObjectRef) {
+			rc = THROW_NPE;
+			goto done;
+		}
+		objectRefClass = J9OBJECT_CLAZZ(_currentThread, originalObjectRef);
+
+		if (!J9_IS_J9CLASS_VALUETYPE(objectRefClass)) {
```
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp]: Can this be ensure by verification instead?

File: runtime/vm/resolvesupport.cpp [hunk 5]
```diff
@@ -939,7 +939,15 @@ resolveInstanceFieldRefInto(J9VMThread *vmStruct, J9Method *method, J9ConstantPo
 					}
 					goto done;
 				}
-				if (!finalFieldSetAllowed(vmStruct, false, method, definingClass, classFromCP, field, canRunJavaCode)) {
+			if (
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+				/* The withfield bytecode is allowed to set a final field. However, the invoker of withfield must have private access
+				 * to the field (similar to a constructor setting a final field) */
+				J9_ARE_NO_BITS_SET(resolveFlags, J9_RESOLVE_FLAG_FIELD_WITHER) &&
```
Comment (gacholio) [inline runtime/vm/resolvesupport.cpp]: This code doesn't seem to match the comment - all withfield of final fields will be allowed.

File: runtime/oti/j9consts.h (hunk 2 above)
Comment (tajila) [reply runtime/oti/j9consts.h]: It a naming convention used in the earlier days of VT. (ie. getter, setter & wither).  
  
Ill change it to withfield for more clarity.

File: runtime/vm/BytecodeInterpreter.hpp (hunk 4 above)
Comment (tajila) [reply runtime/vm/BytecodeInterpreter.hpp]: The verifier will only see the signatures. An 'L' signature may be a VT or just a regular Ref type so we cant do the check there.

File: runtime/vm/resolvesupport.cpp (hunk 5 above)
Comment (tajila) [reply runtime/vm/resolvesupport.cpp]: The private access check is done above. So at this point all that needs to be checked is if it is a with_field.

I will update the comment to make that more clear

File: runtime/vm/BytecodeInterpreter.hpp [hunk 6]
```diff
@@ -8114,8 +8114,91 @@ done:;
...
+				rc = GOTO_THROW_CURRENT_EXCEPTION;
+				goto done;
+			}
+			restoreGenericSpecialStackFrame(REGISTER_ARGS);
+			goto retry;
+		}
+
+		/* Must be done after resolution as we don't know where the object ref is
+		 * until we have resolved the fieldref
+		 */
```
Comment (gacholio) [inline runtime/vm/BytecodeInterpreter.hpp]: This comment appears untrue. It must be done after resolution because resolve errors should take precedence over NPE.

File: runtime/vm/resolvesupport.cpp [hunk 7]
```diff
@@ -741,6 +741,7 @@ resolveStaticFieldRefInto(J9VMThread *vmStruct, J9Method *method, J9ConstantPool
 					}
 					goto done;
 				}
+
```
Comment (gacholio) [inline runtime/vm/resolvesupport.cpp]: Unnecessary change.

File: runtime/vm/BytecodeInterpreter.hpp (hunk 6 above)
Comment (tajila) [reply runtime/vm/BytecodeInterpreter.hpp]: Thats true. What I was trying to say is, we wont know if its a single or double slot field until after resolution so we wont know whether objectref is _sp + 1 or _sp + 2

Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp]: Got it - this would be a lot clearer if the ref weren't fetched before the resolution!

**Comment by tajila** [comment]
@gacholio updated the PR

**Comment by gacholio** [comment]
jenkins test sanity plinux jdk8

**Comment by tajila** [comment]
https://github.com/eclipse/openj9/issues/1390
